// @williambeto/ai-workflow/src/core/healing/cli-remediation-executor.js
import { spawnSync, execSync } from "node:child_process";
import fs from "node:fs/promises";
import path from "node:path";

/**
 * Determines which npm scripts are available in the consumer project.
 * @param {string} cwd
 * @returns {Promise<Record<string, string>>}
 */
async function readScripts(cwd) {
  try {
    const pkg = JSON.parse(await fs.readFile(path.join(cwd, "package.json"), "utf8"));
    return pkg.scripts || {};
  } catch {
    return {};
  }
}

/**
 * Returns the list of files changed since the last commit.
 * Used to detect whether remediation produced any observable change.
 * @param {string} cwd
 * @returns {string[]}
 */
function getChangedFiles(cwd) {
  try {
    const output = execSync("git status --short", { cwd, encoding: "utf8" });
    return output
      .split("\n")
      .filter(Boolean)
      .map((line) => line.slice(3).trim())
      .filter(Boolean);
  } catch {
    return [];
  }
}

/**
 * Maps an affectedCheck source string to likely npm script names to re-run.
 * e.g. "check:branchSafety" → [] (no command can fix a branch issue)
 *      "evidence" with failed test → ["test"]
 * @param {string[]} affectedChecks
 * @param {Record<string, string>} scripts
 * @returns {string[]} ordered list of script names to retry
 */
function resolveRetryScripts(affectedChecks, scripts) {
  const retry = new Set();

  for (const check of affectedChecks) {
    // Evidence failures → re-run the failing validation scripts
    if (check === "evidence" || check === "result") {
      if (scripts.test) retry.add("test");
      if (scripts.lint) retry.add("lint");
      if (scripts.build) retry.add("build");
      if (scripts.typecheck) retry.add("typecheck");
    }
    // Specific check types
    if (check.includes("test")) retry.add("test");
    if (check.includes("lint")) retry.add("lint");
    if (check.includes("build")) retry.add("build");
    if (check.includes("typecheck")) retry.add("typecheck");
  }

  // Filter to only scripts that actually exist
  return [...retry].filter((name) => !!scripts[name]);
}

/**
 * CLI Remediation Executor — default implementation for `ai-workflow run`.
 *
 * Plugs into HealerEngine as the `remediate` function. On each attempt it:
 *   1. Reads the remediation-request.json to understand what failed
 *   2. Re-runs the affected npm scripts (lint, test, build, typecheck)
 *   3. Reports changed files and whether any script now passes
 *
 * Design constraints:
 *   - This executor does NOT modify source code — it only re-runs validation commands.
 *   - It is effective for transient failures (e.g. flaky tests, stale build artifacts).
 *   - For structural failures (missing tests, broken implementation), it will exhaust
 *     its attempt budget and return applied:false, triggering a BLOCKED status.
 *   - The user is shown clear output on each attempt.
 *
 * @param {Object} ctx - Context provided by HealerEngine.run()
 * @param {number} ctx.attempt
 * @param {number} ctx.maxAttempts
 * @param {string} ctx.mode
 * @param {string} ctx.taskSlug
 * @param {Object} ctx.result - Current combined validation result
 * @param {string[]} ctx.affectedChecks
 * @param {string}  ctx.requestPath - Path to remediation-request.json
 * @param {string}  cwd - Consumer project working directory (bound via closure)
 * @returns {Promise<{applied: boolean, changedFiles?: string[], evidenceAdded?: string[], reason?: string}>}
 */
export function createCliRemediationExecutor(cwd) {
  return async function remediationExecutor({ attempt, maxAttempts, mode, affectedChecks, requestPath }) {
    console.log(`\n[HEALER] Attempt ${attempt}/${maxAttempts} — mode: ${mode}`);
    console.log(`[HEALER] Affected: ${affectedChecks.join(", ")}`);

    const scripts = await readScripts(cwd);
    const filesBefore = getChangedFiles(cwd);
    const toRetry = resolveRetryScripts(affectedChecks, scripts);

    if (toRetry.length === 0) {
      console.log("[HEALER] No retryable scripts found for the affected checks.");
      console.log(`[HEALER] Review ${requestPath} for details.`);
      return {
        applied: false,
        reason: `No npm scripts available to address: ${affectedChecks.join(", ")}. Manual remediation required.`
      };
    }

    console.log(`[HEALER] Re-running: ${toRetry.map((s) => `npm run ${s}`).join(", ")}`);

    const results = [];
    let anyPassed = false;

    for (const scriptName of toRetry) {
      process.stdout.write(`  → npm run ${scriptName} ... `);
      const result = spawnSync("npm", ["run", scriptName], {
        cwd,
        shell: true,
        encoding: "utf8",
        timeout: 120000
      });
      const passed = result.status === 0;
      if (passed) anyPassed = true;
      console.log(passed ? "PASS" : "FAIL");
      results.push({ script: scriptName, passed, exitCode: result.status });
    }

    const filesAfter = getChangedFiles(cwd);
    const changedFiles = filesAfter.filter((f) => !filesBefore.includes(f));

    if (!anyPassed && changedFiles.length === 0) {
      console.log("[HEALER] No scripts passed and no files changed — remediation had no effect.");
      return {
        applied: false,
        reason: `Re-ran ${toRetry.join(", ")} — all failed. Structural fix required by the implementing agent.`
      };
    }

    console.log(`[HEALER] Applied: ${results.filter((r) => r.passed).map((r) => r.script).join(", ") || "none passed"}`);
    if (changedFiles.length) console.log(`[HEALER] Changed files: ${changedFiles.join(", ")}`);

    return {
      applied: true,
      changedFiles,
      evidenceAdded: anyPassed ? ["validation-output"] : []
    };
  };
}
