// @williambeto/ai-workflow/src/core/healing/healer-engine.js
import fs from "node:fs/promises";
import path from "node:path";
import crypto from "node:crypto";
import { isRecoverableGateFailure } from "../statuses.js";

const MODE_ATTEMPTS = Object.freeze({ quick: 1, standard: 2, full: 3 });

function stableJson(value) {
  if (Array.isArray(value)) return `[${value.map(stableJson).join(",")}]`;
  if (value && typeof value === "object") {
    return `{${Object.keys(value).sort().map((key) => `${JSON.stringify(key)}:${stableJson(value[key])}`).join(",")}}`;
  }
  return JSON.stringify(value);
}

function collectBlockingFindings(result = {}) {
  const findings = [];
  const add = (source, value) => {
    if (!value || typeof value !== "object") return;
    const status = value.status || value.overallStatus;
    if (!["FAIL", "FAIL_QUALITY_GATE", "FAIL_DELEGATION_GATE", "BLOCKED"].includes(status)) return;
    findings.push({
      source,
      status,
      missing: Array.isArray(value.missing) ? [...value.missing].sort() : [],
      reason: value.reason || value.summary || ""
    });
  };

  add("result", result);
  add("evidence", result.evidence);
  add("quality", result.quality);
  const checks = result.quality?.checks || result.policyValidation?.checks || result.checks || {};
  for (const [name, check] of Object.entries(checks)) add(`check:${name}`, check);
  return findings.sort((a, b) => a.source.localeCompare(b.source));
}

export class HealerEngine {
  constructor({ cwd, mode = "standard", taskSlug = "implementation", maxAttempts = null } = {}) {
    this.cwd = cwd;
    this.mode = MODE_ATTEMPTS[mode] ? mode : "standard";
    this.taskSlug = taskSlug;
    this.maxAttempts = maxAttempts ?? MODE_ATTEMPTS[this.mode];
  }

  get statePath() {
    return path.join(this.cwd, ".ai-workflow", `healing-state-${this.taskSlug}.json`);
  }

  fingerprint(result) {
    return crypto.createHash("sha256").update(stableJson(collectBlockingFindings(result))).digest("hex");
  }

  blockingCount(result) {
    return collectBlockingFindings(result).length;
  }

  affectedChecks(result) {
    return collectBlockingFindings(result).map((finding) => finding.source);
  }

  hasProgress(before, after, remediation = {}) {
    if (!isRecoverableGateFailure(after.overallStatus)) return true;
    const beforeCount = this.blockingCount(before);
    const afterCount = this.blockingCount(after);
    if (afterCount < beforeCount) return true;
    if (this.fingerprint(before) !== this.fingerprint(after) && (remediation.changedFiles?.length || remediation.evidenceAdded?.length)) return true;
    return false;
  }

  async writeState(state) {
    await fs.mkdir(path.dirname(this.statePath), { recursive: true });
    await fs.writeFile(this.statePath, JSON.stringify(state, null, 2));
  }

  async resetState() {
    try { await fs.unlink(this.statePath); } catch {}
  }

  async writeRemediationRequest({ attempt, result }) {
    const requestPath = path.join(this.cwd, ".ai-workflow", "remediation-request.json");
    await fs.mkdir(path.dirname(requestPath), { recursive: true });
    const request = {
      taskSlug: this.taskSlug,
      executionMode: this.mode,
      attempt,
      maxAttempts: this.maxAttempts,
      status: result.overallStatus,
      affectedChecks: this.affectedChecks(result),
      findingsFingerprint: this.fingerprint(result),
      instruction: "Apply the smallest safe correction for the listed gates, then rerun validation. Do not broaden scope."
    };
    await fs.writeFile(requestPath, JSON.stringify(request, null, 2));
    return requestPath;
  }

  async run({ initialResult, validate, remediate }) {
    let current = initialResult;
    const history = [];
    const seen = new Set();

    if (!isRecoverableGateFailure(current.overallStatus)) {
      await this.resetState();
      return { ...current, remediation: { status: "NOT_REQUIRED", attempts: 0, history } };
    }

    for (let attempt = 1; attempt <= this.maxAttempts; attempt += 1) {
      const fingerprint = this.fingerprint(current);
      if (seen.has(fingerprint)) {
        const blocked = { overallStatus: "BLOCKED", reason: "Repeated unresolved finding", previousResult: current };
        await this.writeState({ status: "BLOCKED", reason: blocked.reason, attempts: attempt - 1, history });
        return { ...blocked, remediation: { status: "BLOCKED", attempts: attempt - 1, history } };
      }
      seen.add(fingerprint);

      const requestPath = await this.writeRemediationRequest({ attempt, result: current });
      if (typeof remediate !== "function") {
        const blocked = { overallStatus: "BLOCKED", reason: "No remediation executor available", previousResult: current, requestPath };
        await this.writeState({ status: "BLOCKED", reason: blocked.reason, attempts: attempt - 1, history, requestPath });
        return { ...blocked, remediation: { status: "BLOCKED", attempts: attempt - 1, history, requestPath } };
      }

      const remediation = await remediate({
        attempt,
        maxAttempts: this.maxAttempts,
        mode: this.mode,
        taskSlug: this.taskSlug,
        result: current,
        affectedChecks: this.affectedChecks(current),
        requestPath
      });

      if (!remediation?.applied) {
        const blocked = { overallStatus: "BLOCKED", reason: remediation?.reason || "Remediation was not applied", previousResult: current, requestPath };
        history.push({ attempt, before: current.overallStatus, remediation, after: "NOT_RUN", progress: false });
        await this.writeState({ status: "BLOCKED", reason: blocked.reason, attempts: attempt, history });
        return { ...blocked, remediation: { status: "BLOCKED", attempts: attempt, history } };
      }

      const next = await validate({ affectedChecks: this.affectedChecks(current), attempt });
      const progress = this.hasProgress(current, next, remediation);
      history.push({
        attempt,
        before: current.overallStatus,
        after: next.overallStatus,
        progress,
        changedFiles: remediation.changedFiles || [],
        evidenceAdded: remediation.evidenceAdded || [],
        affectedChecks: this.affectedChecks(current)
      });
      await this.writeState({ status: next.overallStatus, attempts: attempt, history });

      if (!isRecoverableGateFailure(next.overallStatus)) {
        await this.resetState();
        return { ...next, remediation: { status: "RESOLVED", attempts: attempt, history } };
      }

      if (!progress) {
        return {
          overallStatus: "BLOCKED",
          reason: "Remediation made no measurable progress",
          previousResult: next,
          remediation: { status: "BLOCKED", attempts: attempt, history }
        };
      }

      current = next;
    }

    return {
      overallStatus: "BLOCKED",
      reason: `Quality remediation attempts exhausted for ${this.mode} mode`,
      previousResult: current,
      remediation: { status: "BLOCKED", attempts: this.maxAttempts, history }
    };
  }
}

export { MODE_ATTEMPTS, collectBlockingFindings };
