// @williambeto/ai-workflow/src/core/package-assets.js
/**
 * Package asset resolution for @williambeto/ai-workflow.
 *
 * Resolves the installed package root and reads real skill/agent/command
 * files so `init` copies them to the target project instead of generating stubs.
 */
import { fileURLToPath } from "node:url";
import { dirname, resolve } from "node:path";
import { readFileSync, readdirSync, statSync } from "node:fs";

let cachedRoot = null;

/**
 * Return the absolute path to the installed package root
 * (the directory containing `.agents/skills/`).
 *
 * It dynamically searches for the package.json belonging to @williambeto/ai-workflow
 * by climbing up from the current module location.
 */
export function getPackageRoot() {
  if (cachedRoot) return cachedRoot;
  let currentDir = dirname(fileURLToPath(import.meta.url));

  while (true) {
    const pkgPath = resolve(currentDir, "package.json");
    try {
      const pkg = JSON.parse(readFileSync(pkgPath, "utf8"));
      if (pkg.name === "@williambeto/ai-workflow") {
        cachedRoot = currentDir;
        return cachedRoot;
      }
    } catch {
      // package.json doesn't exist or is invalid, keep climbing
    }

    const parentDir = dirname(currentDir);
    if (parentDir === currentDir) {
      break; // Reached filesystem root
    }
    currentDir = parentDir;
  }

  throw new Error("Failed to resolve @williambeto/ai-workflow package root physically.");
}

/**
 * Read a file relative to the package root.
 * Returns null if the file does not exist.
 */
export function readPackageFile(relativePath) {
  try {
    const fullPath = resolve(getPackageRoot(), relativePath);
    return readFileSync(fullPath, "utf8");
  } catch {
    return null;
  }
}

/**
 * Resolve the current kit version from package metadata.
 * Returns null if package.json cannot be read/parsed.
 */
export function getPackageVersion() {
  try {
    const raw = readPackageFile("package.json");
    if (!raw) return null;
    const parsed = JSON.parse(raw);
    return typeof parsed.version === "string" ? parsed.version : null;
  } catch {
    return null;
  }
}

/**
 * Discover ALL files under a package-relative directory.
 * Returns a flat map of { relativePath: content }.
 * Skips non-existent directories silently.
 */
export function discoverPackageFiles(baseRelativePath) {
  const files = {};
  const basePath = resolve(getPackageRoot(), baseRelativePath);

  try {
    if (!statSync(basePath).isDirectory()) {
      return files;
    }
  } catch {
    return files;
  }

  function walk(dir, relativePrefix) {
    const entries = readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      const fullPath = resolve(dir, entry.name);
      const relativePath = relativePrefix
        ? `${relativePrefix}/${entry.name}`
        : entry.name;

      if (entry.isDirectory()) {
        walk(fullPath, relativePath);
      } else if (entry.isFile()) {
        try {
          files[relativePath] = readFileSync(fullPath, "utf8");
        } catch {
          // skip unreadable files
        }
      }
    }
  }

  walk(basePath, baseRelativePath);
  return files;
}

/**
 * Read the SKILL.md for a given skill name, plus any reference files
 * found under a directory.
 * Returns a map of { relativePath: content }.
 */
export function getFullSkillFiles(skillName) {
  const skillRelPath = `dist-assets/skills/${skillName}`;
  const files = discoverPackageFiles(skillRelPath);
  return files;
}

/**
 * Read the agent .md file for a given agent name.
 * Returns the content string or null.
 */
export function getFullAgentContent(agentName) {
  return readPackageFile(`dist-assets/agents/${agentName}.md`);
}

/**
 * Read the command .md file for a given command name.
 * Returns the content string or null.
 */
export function getFullCommandContent(commandName) {
  return readPackageFile(`dist-assets/commands/${commandName}.md`);
}
