// @williambeto/ai-workflow/src/core/workflow-profiles.js
const PROFILE_DEFINITIONS = Object.freeze({
  "frontend-product": Object.freeze({
    owner: "Astra",
    skills: ["frontend-development", "ui-ux-design"],
    objective: "Deliver a user-facing product or marketing surface with truthful copy and deliberate visual composition.",
    requiredChecks: ["tests", "typecheck", "build", "responsive-render", "primary-interaction", "truthfulness"],
    forbiddenAssumptions: ["fixed SaaS section sequence", "mandatory pricing", "mandatory testimonials", "subjective premium score"]
  }),
  "frontend-utility": Object.freeze({
    owner: "Astra",
    skills: ["frontend-development", "ui-ux-design"],
    objective: "Deliver a focused user tool with a complete primary flow and clear operational states.",
    requiredChecks: ["tests", "typecheck", "build", "responsive-render", "primary-interaction"],
    forbiddenAssumptions: ["marketing sections", "pricing", "testimonials", "commercial narrative"]
  }),
  "backend-api": Object.freeze({
    owner: "Astra",
    skills: ["backend-development"],
    objective: "Deliver a bounded backend/API change with validated contracts, errors, authorization, and tests.",
    requiredChecks: ["tests", "build-or-typecheck", "input-validation", "failure-paths"],
    forbiddenAssumptions: ["frontend deliverables", "visual evidence"]
  }),
  refactor: Object.freeze({
    owner: "Astra",
    skills: ["refactoring"],
    objective: "Improve structure without changing observable behavior outside the approved scope.",
    requiredChecks: ["behavior-preservation", "tests", "focused-diff"],
    forbiddenAssumptions: ["new product behavior", "unrelated cleanup"]
  }),
  documentation: Object.freeze({
    owner: "Astra",
    skills: ["documentation"],
    objective: "Create or update accurate documentation grounded in the current repository behavior.",
    requiredChecks: ["references", "paths", "commands", "consistency"],
    forbiddenAssumptions: ["implementation claims without evidence"]
  }),
  "security-review": Object.freeze({
    owner: "Sage",
    skills: ["qa-workflow", "technical-leadership"],
    objective: "Inspect security-relevant behavior and report evidence-ranked findings without implementing unrelated changes.",
    requiredChecks: ["evidence", "severity", "runtime-vs-dev", "remediation-owner"],
    forbiddenAssumptions: ["automatic force fixes", "unverified exploitability"]
  }),
  generic: Object.freeze({
    owner: "Atlas",
    skills: [],
    objective: "Route work without inventing domain requirements.",
    requiredChecks: ["scope", "owner", "mode"],
    forbiddenAssumptions: ["domain-specific structure"]
  })
});

/**
 * NOTE: requiredChecks are advisory, not programmatically enforced.
 *
 * Each profile's `requiredChecks` array is informational — it guides agent behaviour,
 * informs the handoff document, and communicates domain expectations to the AI.
 * QualityGuard and EvidenceCollector do NOT iterate requiredChecks to block delivery.
 *
 * The rationale: profile-specific checks (e.g. "responsive-render", "primary-interaction")
 * are context-dependent and require human or agent judgement. Automated enforcement
 * would require domain-specific validators that don't currently exist.
 *
 * If you add automated enforcement for a profile check, update this comment.
 */

const PROFILE_PATTERNS = Object.freeze([
  ["security-review", /\b(?:security review|security audit|vulnerability audit|threat model)\b/i],
  ["refactor", /\b(?:refactor|restructure|cleanup without behavior change|preserve behavior)\b/i],
  ["documentation", /\b(?:documentation only|write docs|update readme|document the|documentation task)\b/i],
  ["backend-api", /\b(?:api endpoint|backend api|rest api|graphql|controller|service endpoint|authenticated api)\b/i],
  ["frontend-product", /\b(?:landing page|marketing page|product page|pricing page|homepage|redesign|brand page)\b/i],
  ["frontend-utility", /\b(?:validator|search page|lookup|dashboard|form|calculator|tool|admin page|repository search)\b/i]
]);

export function getWorkflowProfile(name = "generic") {
  return PROFILE_DEFINITIONS[name] || PROFILE_DEFINITIONS.generic;
}

export function listWorkflowProfiles() {
  return Object.keys(PROFILE_DEFINITIONS);
}

export function resolveWorkflowProfile({ request = "", explicitProfile = null } = {}) {
  if (explicitProfile && PROFILE_DEFINITIONS[explicitProfile]) return explicitProfile;
  const text = String(request).trim();
  for (const [profile, pattern] of PROFILE_PATTERNS) {
    if (pattern.test(text)) return profile;
  }
  return "generic";
}

export function buildCompactProfileHandoff({ profile, objective, requirements = [], constraints = [] }) {
  const selected = getWorkflowProfile(profile);
  const unique = (values) => [...new Set(values.filter(Boolean))];
  return {
    profile,
    owner: selected.owner,
    skills: [...selected.skills],
    objective: objective || selected.objective,
    requirements: unique(requirements),
    constraints: unique(constraints),
    requiredChecks: [...selected.requiredChecks]
  };
}

export { PROFILE_DEFINITIONS };
