// @williambeto/ai-workflow/src/core/validation/evidence-collector.js
import { spawnSync } from "node:child_process";
import fs from "node:fs/promises";
import path from "node:path";
import { QualityGuard } from "./quality-guard.js";

const VALIDATION_KINDS = new Set(["test", "build", "typecheck", "lint", "security", "smoke", "validate", "other"]);

function publicStatus(status) {
  if (status === "PASS") return "COMPLETED";
  if (status === "PASS_WITH_NOTES") return "COMPLETED_WITH_NOTES";
  return "BLOCKED";
}

function inferTaskKind(task = {}) {
  const explicit = String(task.kind || "").trim().toLowerCase();
  if (VALIDATION_KINDS.has(explicit)) return explicit;

  const value = `${task.name || ""} ${task.command || ""}`.toLowerCase();
  if (/\b(test|tests|vitest|jest|mocha|ava|tap|pytest|phpunit|rspec|cypress|playwright)\b/.test(value)) return "test";
  if (/\b(typecheck|tsc\b|mypy|pyright)\b/.test(value)) return "typecheck";
  if (/\b(lint|eslint|stylelint|ruff|flake8|pylint)\b/.test(value)) return "lint";
  if (/\b(build|bundle|compile)\b/.test(value)) return "build";
  if (/\b(audit|security|sast|scan)\b/.test(value)) return "security";
  if (/\b(smoke|curl|preview|serve)\b/.test(value)) return "smoke";
  if (/\bvalidate\b/.test(value)) return "validate";
  return "other";
}

export class EvidenceCollector {
  constructor({ cwd, maxLogLength = 2000, timeout = 60000, taskSlug = null, mode = null, profile = "generic", branchRecovery = "NOT_RECORDED" } = {}) {
    this.cwd = cwd;
    this.maxLogLength = maxLogLength;
    this.timeout = timeout;
    this.taskSlug = taskSlug;
    this.mode = mode;
    this.profile = profile;
    this.branchRecovery = branchRecovery;
  }

  runTask(task) {
    const kind = inferTaskKind(task);
    if (task.presetStatus) {
      return { name: task.name, command: task.command, kind, status: task.presetStatus, exitCode: null, summary: task.summary, output: "" };
    }
    const result = spawnSync(task.command, {
      cwd: this.cwd,
      shell: true,
      encoding: "utf8",
      timeout: this.timeout
    });
    let status = result.status === 0 ? "PASS" : "FAIL";
    let summary = status === "PASS" ? "Command completed successfully." : "Command failed.";
    if (result.error?.code === "ETIMEDOUT") {
      status = "BLOCKED";
      summary = `Command timed out after ${this.timeout / 1000}s.`;
    }
    const rawOutput = `${result.stdout || ""}${result.stderr || ""}`.trim();
    const output = rawOutput.length > this.maxLogLength ? `${rawOutput.slice(0, this.maxLogLength)}\n... [TRUNCATED]` : rawOutput;
    return { name: task.name, command: task.command, kind, status, exitCode: result.status, signal: result.signal, summary, output };
  }

  async collect(tasks, { writeArtifact = false } = {}) {
    const qualityGuard = new QualityGuard({ cwd: this.cwd, taskSlug: this.taskSlug, mode: this.mode, evidenceWillBePersisted: writeArtifact });
    const implementation = qualityGuard.isImplementationTask();
    const executableBehavior = await qualityGuard.hasExecutableBehaviorChanges();
    const results = tasks.map((task) => this.runTask(task));
    const policyValidation = await qualityGuard.verify();
    const behaviorTests = results.filter((result) => result.kind === "test");
    const passingBehaviorTest = behaviorTests.some((result) => result.status === "PASS");

    let overallStatus = "PASS";
    if (implementation && executableBehavior && tasks.length === 0) overallStatus = "BLOCKED";
    else if (results.some((result) => ["FAIL", "BLOCKED", "FAIL_QUALITY_GATE"].includes(result.status))) overallStatus = "FAIL_QUALITY_GATE";
    else if (executableBehavior && !passingBehaviorTest) overallStatus = "BLOCKED";
    else if (["FAIL", "BLOCKED", "FAIL_QUALITY_GATE"].includes(policyValidation.overallStatus)) overallStatus = "FAIL_QUALITY_GATE";
    else if (results.some((result) => result.status === "PASS_WITH_NOTES") || policyValidation.overallStatus === "PASS_WITH_NOTES") overallStatus = "PASS_WITH_NOTES";

    const limitations = [];
    if (implementation && executableBehavior && tasks.length === 0) limitations.push("No meaningful validation command was available for executable implementation work.");
    if (executableBehavior && behaviorTests.length === 0) {
      limitations.push("Executable behavior changed without a proportional automated behavior test. Build, lint, typecheck, smoke, screenshots, and manual review do not replace behavior tests.");
    }
    for (const [name, check] of Object.entries(policyValidation.checks || {})) {
      if (check.status === "PASS_WITH_NOTES" && check.reason) limitations.push(`${name}: ${check.reason}`);
    }

    const evidence = {
      timestamp: new Date().toISOString(),
      taskSlug: this.taskSlug,
      executionMode: this.mode,
      workflowProfile: this.profile,
      branchRecovery: this.branchRecovery,
      branch: policyValidation.git?.branch || "unknown",
      changedFiles: policyValidation.git?.changedFiles || [],
      status: publicStatus(overallStatus),
      internalStatus: overallStatus,
      commands: results,
      checks: policyValidation.checks,
      limitations
    };

    if (writeArtifact) {
      await fs.writeFile(path.join(this.cwd, "EVIDENCE.json"), JSON.stringify(evidence, null, 2));
    }
    return evidence;
  }
}

export { inferTaskKind };
