// @williambeto/ai-workflow/src/core/validation/quality-guard.js
import { execSync } from "node:child_process";
import fs from "node:fs/promises";
import path from "node:path";

const PROTECTED_BRANCHES = new Set(["main", "master"]);

function normalizePath(file) {
  return String(file || "").trim().replaceAll("\\", "/");
}

function parseStatusLine(line) {
  const value = String(line || "");
  const match = value.match(/^(?:[ MADRCU?!]{2}\s+|[MADRCU?!]\s+)(.+)$/);
  if (!match) return "";
  const candidate = match[1].trim();
  const renamed = candidate.includes(" -> ") ? candidate.split(" -> ").at(-1) : candidate;
  return normalizePath(renamed);
}

function isManagedChangePath(file = "") {
  return /(^|\/)(node_modules|vendor|dist|coverage|\.cache)(\/|$)/.test(normalizePath(file));
}

async function exists(target) {
  try {
    await fs.access(target);
    return true;
  } catch {
    return false;
  }
}

export { parseStatusLine };

/**
 * QualityGuard verifies objective delivery safety. It does not grade prose or
 * require ceremonial workflow files.
 */
export class QualityGuard {
  constructor({ cwd, taskSlug = null, mode = null, evidenceWillBePersisted = false } = {}) {
    this.cwd = cwd;
    this.taskSlug = taskSlug;
    this.mode = mode;
    this._changedFiles = null;
    this._hasHead = null;
    this.evidenceWillBePersisted = evidenceWillBePersisted;
  }

  runGit(command) {
    return execSync(command, {
      cwd: this.cwd,
      encoding: "utf8",
      stdio: ["ignore", "pipe", "pipe"]
    }).trim();
  }

  hasGitHead() {
    if (this._hasHead !== null) return this._hasHead;
    try {
      this.runGit("git rev-parse --verify HEAD");
      this._hasHead = true;
    } catch {
      this._hasHead = false;
    }
    return this._hasHead;
  }

  getCurrentBranch() {
    try {
      return this.runGit("git branch --show-current") || "unknown";
    } catch {
      return "unknown";
    }
  }

  getChangedFiles() {
    if (this._changedFiles) return this._changedFiles;
    const files = new Set();
    try {
      if (this.hasGitHead()) {
        this.runGit("git diff --name-only HEAD")
          .split("\n")
          .map(normalizePath)
          .filter(Boolean)
          .forEach((file) => files.add(file));
      }
      this.runGit("git status --short --untracked-files=all")
        .split("\n")
        .map(parseStatusLine)
        .filter(Boolean)
        .filter((file) => !isManagedChangePath(file))
        .forEach((file) => files.add(file));
    } catch {
      // A non-git consumer may still run validation commands. Branch safety is
      // reported as unavailable rather than fabricated.
    }
    this._changedFiles = [...files].sort();
    return this._changedFiles;
  }

  isDocsTask() {
    const files = this.getChangedFiles();
    return files.length > 0 && files.every((file) => file.endsWith(".md") || file.endsWith(".txt") || file.startsWith("docs/"));
  }

  isUiTask() {
    return this.getChangedFiles().some((file) =>
      /\.(tsx|jsx|vue|html|css|scss|less|blade\.php)$/.test(file) ||
      /(^|\/)(components|pages|views)\//.test(file)
    );
  }

  isImplementationTask() {
    const files = this.getChangedFiles();
    if (files.length === 0 || this.isDocsTask()) return false;
    return files.some((file) =>
      file === "package.json" ||
      /^(src|app|pages|components|server|api)\//.test(file) ||
      /\.(js|jsx|ts|tsx|vue|php|py|rb|go|java|css|scss|html)$/.test(file)
    );
  }


  async hasExecutableBehaviorChanges() {
    const files = this.getChangedFiles().filter((file) => !isManagedChangePath(file));
    const executableExtensions = /\.(js|jsx|mjs|cjs|ts|tsx|vue|php|py|rb|go|java|cs|rs|swift|kt|kts|sh|bash)$/i;
    if (files.some((file) => executableExtensions.test(file))) return true;

    for (const file of files.filter((item) => /\.(html?|xhtml)$/i.test(item))) {
      try {
        const content = await fs.readFile(path.join(this.cwd, file), "utf8");
        if (/<script\b|\son(?:click|change|input|submit|keydown|keyup|load|error)\s*=|javascript:/i.test(content)) return true;
      } catch {
        // Missing or unreadable changed files are handled by the normal validation path.
      }
    }
    return false;
  }

  verifyBranchSafety() {
    const branch = this.getCurrentBranch();
    const implementation = this.isImplementationTask();
    if (!this.hasGitHead()) {
      return { status: "PASS_WITH_NOTES", branch, reason: "Git HEAD unavailable; branch safety could not be verified." };
    }
    if (implementation && PROTECTED_BRANCHES.has(branch)) {
      return { status: "FAIL_QUALITY_GATE", branch, reason: `Implementation changes are present on protected branch '${branch}'.` };
    }
    return { status: "PASS", branch };
  }

  async verifyDocumentationNeed() {
    if (!this.isImplementationTask()) return { status: "NOT_RUN", reason: "No implementation changes detected." };
    const files = this.getChangedFiles();
    const docsChanged = files.some((file) => file === "README.md" || file.startsWith("docs/") || file.endsWith(".md"));
    const publicSurfaceChanged = files.some((file) =>
      file === "package.json" ||
      /(^|\/)(routes?|api|cli|commands?|config|public)\//.test(file) ||
      /README|CHANGELOG/.test(file)
    );
    if (publicSurfaceChanged && !docsChanged) {
      return { status: "PASS_WITH_NOTES", reason: "Public behavior may have changed without a documentation update." };
    }
    return { status: "PASS" };
  }

  async verifyUiEvidence() {
    if (!this.isUiTask()) return { status: "NOT_RUN", reason: "Non-UI task." };
    const candidates = ["docs", "screenshots", "evidence"];
    const found = [];
    for (const dir of candidates) {
      const root = path.join(this.cwd, dir);
      if (!(await exists(root))) continue;
      const entries = await fs.readdir(root, { recursive: true }).catch(() => []);
      for (const entry of entries) {
        if (/\.(png|jpe?g|webp)$/i.test(String(entry))) found.push(normalizePath(path.join(dir, String(entry))));
      }
    }
    return found.length
      ? { status: "PASS", screenshots: found.slice(0, 20) }
      : { status: "PASS_WITH_NOTES", reason: "UI changed without persisted screenshots; rely on the stated browser/manual review only if it actually occurred." };
  }

  async verifyFullModeEvidence() {
    if (this.mode !== "full") return { status: "NOT_RUN", reason: "Persisted evidence is optional outside full mode." };
    if (this.evidenceWillBePersisted) return { status: "PASS", reason: "Collector will persist observed full-mode evidence." };
    const candidates = [
      path.join(this.cwd, "EVIDENCE.json"),
      this.taskSlug ? path.join(this.cwd, "docs/workflows", this.taskSlug, "evidence.json") : null
    ].filter(Boolean);
    for (const file of candidates) {
      if (await exists(file)) return { status: "PASS", file: path.relative(this.cwd, file) };
    }
    return { status: "FAIL_QUALITY_GATE", reason: "Full mode requires persisted evidence generated from observed execution." };
  }

  verifyTruthfulnessBoundary() {
    if (!this.isUiTask()) return { status: "NOT_RUN", reason: "Non-UI task." };
    return {
      status: "NOT_RUN",
      reason: "Product truthfulness requires semantic review; keyword matching is intentionally not used as proof."
    };
  }

  async verify() {
    const branchSafety = this.verifyBranchSafety();
    const documentation = await this.verifyDocumentationNeed();
    const uiEvidence = await this.verifyUiEvidence();
    const fullModeEvidence = await this.verifyFullModeEvidence();
    const productTruthfulness = this.verifyTruthfulnessBoundary();

    // productTruthfulness is intentionally excluded from the blocking checks array.
    // Semantic truthfulness cannot be verified by keyword matching; it always returns
    // NOT_RUN by design. It appears in the output as an observational signal for
    // human reviewers, but never affects overallStatus.
    const checks = [branchSafety, documentation, uiEvidence, fullModeEvidence];
    let overallStatus = "PASS";
    if (checks.some((check) => ["FAIL", "FAIL_QUALITY_GATE", "BLOCKED"].includes(check.status))) {
      overallStatus = "FAIL_QUALITY_GATE";
    } else if (checks.some((check) => check.status === "PASS_WITH_NOTES")) {
      overallStatus = "PASS_WITH_NOTES";
    }

    return {
      overallStatus,
      isUiTask: this.isUiTask(),
      isDocsTask: this.isDocsTask(),
      isImplementationTask: this.isImplementationTask(),
      git: {
        hasHead: this.hasGitHead(),
        branch: this.getCurrentBranch(),
        changedFiles: this.getChangedFiles()
      },
      checks: {
        branchSafety,
        documentation,
        uiEvidence,
        fullModeEvidence,
        productTruthfulness
      }
    };
  }
}
