// @williambeto/ai-workflow/src/cli.js
import { getPackageVersion } from "./core/package-assets.js";
import { runInit } from "./commands/init.js";
import { runDoctor } from "./commands/doctor.js";
import { runCollectEvidence } from "./commands/collect-evidence.js";
import { runMasterOrchestrator } from "./commands/run.js";

function printHelp() {
  console.log(`ai-workflow

Usage:
  ai-workflow run --spec-path=<path>
  ai-workflow init [--yes] [--force] [--dry-run] [--no-install] [--no-overwrite] [--gemini] [--claude] [--codex] [--profile=<profile>]
  ai-workflow collect-evidence [--task=<slug>] [--mode=<quick|standard|full>] [--dry-run]
  ai-workflow doctor

Commands:
  run               Proportionate orchestrator with branch safety, validation, and bounded remediation
  init              Install AI workflow defaults (OpenCode). --profile=standard (default) or --profile=full (adds examples)
  collect-evidence  Run observed project validation; persists EVIDENCE.json only for full mode
  doctor            Verify local ai-workflow installation
`);
}

function parseFlags(args) {
  const specPathArg = args.find((arg) => arg.startsWith("--spec-path="));
  const profileEqArg = args.find((arg) => arg.startsWith("--profile="));
  const taskArg = args.find((arg) => arg.startsWith("--task="));
  const modeArg = args.find((arg) => arg.startsWith("--mode="));
  const profileIdx = args.indexOf("--profile");
  const profileVal = profileEqArg
    ? profileEqArg.replace("--profile=", "")
    : profileIdx >= 0 && profileIdx + 1 < args.length && !args[profileIdx + 1].startsWith("--")
      ? args[profileIdx + 1]
      : undefined;

  return {
    yes: args.includes("--yes"),
    force: args.includes("--force"),
    dryRun: args.includes("--dry-run"),
    noInstall: args.includes("--no-install"),
    noOverwrite: args.includes("--no-overwrite"),
    gemini: args.includes("--gemini"),
    claude: args.includes("--claude"),
    codex: args.includes("--codex"),
    "dev-mode": args.includes("--dev-mode"),
    specPath: specPathArg ? specPathArg.replace("--spec-path=", "") : undefined,
    profile: profileVal || undefined,
    taskSlug: taskArg ? taskArg.replace("--task=", "") : undefined,
    mode: modeArg ? modeArg.replace("--mode=", "") : undefined
  };
}


export async function runCli(args) {
  const [command] = args;

  if (command === "--version" || command === "-v") {
    console.log(getPackageVersion() || "unknown");
    return;
  }

  if (!command || command === "--help" || command === "-h") {
    printHelp();
    return;
  }

  if (command === "run") {
    await runMasterOrchestrator({
      cwd: process.cwd(),
      ...parseFlags(args.slice(1))
    });
    return;
  }

  if (command === "init") {
    await runInit({
      cwd: process.cwd(),
      ...parseFlags(args.slice(1))
    });
    return;
  }

  if (command === "collect-evidence") {
    const flags = parseFlags(args.slice(1));
    if (flags.mode && !["quick", "standard", "full"].includes(flags.mode)) throw new Error("--mode must be quick, standard, or full");
    await runCollectEvidence({ cwd: process.cwd(), taskSlug: flags.taskSlug, mode: flags.mode, dryRun: flags.dryRun });
    return;
  }

  if (command === "doctor") {
    await runDoctor({ cwd: process.cwd() });
    return;
  }

  throw new Error(`unknown command: ${command}`);
}
