// @williambeto/ai-workflow/src/adapters/platforms/codex.js
import fs from "node:fs/promises";
import path from "node:path";
import { COMPLETION_STATUS_TEXT } from "../../core/statuses.js";

/**
 * Codex Adapter - Transforms OpenCode agents, skills, and commands into Codex-native formats.
 */
export class CodexAdapter {
  constructor({ cwd }) {
    this.cwd = cwd;
  }

  /**
   * Transforms an OpenCode agent into a Codex prompt with completion contract.
   */
  async transformAgent(filePath) {
    const content = await fs.readFile(filePath, "utf8");
    const fileName = path.basename(filePath, ".md");
    
    const hardenedContent = `${content}

##  Completion Contract (Mandatory)
Every task completion MUST provide the following payload:
1. **Status**: ${COMPLETION_STATUS_TEXT}
2. **Scope reviewed**: Brief summary of what was analyzed.
3. **Findings**: Evidence-based discoveries.
4. **Files affected**: List of all concrete paths modified or created.
5. **Validation/evidence**: Commands run and their results.
6. **Risks/notes**: Any caveats or technical debt introduced.
7. **Required manager action**: Clear next step for the orchestrator.
8. **Can manager continue**: [yes/no]
`;

    return {
      content: hardenedContent,
      name: fileName
    };
  }

  /**
   * Deploys agents, skills, and commands to Codex-native directories.
   * Mapping:
   * - Agents -> .github/agents/
   * - Skills -> .agents/skills/
   * - Commands -> .codex/prompts/
   */
  async deploy(installRoot = ".ai-workflow") {
    const sourceAgentsDir = path.join(this.cwd, installRoot, "opencode/agents");
    const sourceSkillsDir = path.join(this.cwd, installRoot, "opencode/skills");
    const sourceCommandsDir = path.join(this.cwd, installRoot, "opencode/commands");

    // 1. Deploy Agents to .github/agents/
    const codexAgentsDir = path.join(this.cwd, ".github/agents");
    await fs.mkdir(codexAgentsDir, { recursive: true });
    
    const agentFiles = await fs.readdir(sourceAgentsDir).catch(() => []);
    for (const file of agentFiles) {
      if (!file.endsWith(".md")) continue;
      const transformed = await this.transformAgent(path.join(sourceAgentsDir, file));
      await fs.writeFile(path.join(codexAgentsDir, `${transformed.name}.md`), transformed.content);
    }

    // 2. Deploy Skills to .agents/skills/
    const codexSkillsDir = path.join(this.cwd, ".agents/skills");
    await fs.mkdir(codexSkillsDir, { recursive: true });

    const skillFolders = await fs.readdir(sourceSkillsDir).catch(() => []);
    for (const folder of skillFolders) {
      const skillSource = path.join(sourceSkillsDir, folder, "SKILL.md");
      if (await this.exists(skillSource)) {
        const targetDir = path.join(codexSkillsDir, folder);
        await fs.mkdir(targetDir, { recursive: true });
        const content = await fs.readFile(skillSource, "utf8");
        await fs.writeFile(path.join(targetDir, "SKILL.md"), content);
      }
    }

    // 3. Deploy Commands to .codex/prompts/
    const codexPromptsDir = path.join(this.cwd, ".codex/prompts");
    await fs.mkdir(codexPromptsDir, { recursive: true });

    const commandFiles = await fs.readdir(sourceCommandsDir).catch(() => []);
    for (const file of commandFiles) {
      if (!file.endsWith(".md") && !file.endsWith(".toml")) continue;
      // Codex prompts work best as .md
      const targetFileName = file.endsWith(".toml") ? `${path.basename(file, ".toml")}.md` : file;
      const content = await fs.readFile(path.join(sourceCommandsDir, file), "utf8");
      await fs.writeFile(path.join(codexPromptsDir, targetFileName), content);
    }
  }

  async exists(p) {
    try {
      await fs.access(p);
      return true;
    } catch {
      return false;
    }
  }
}
