// @williambeto/ai-workflow/src/adapters/platforms/gemini.js
import fs from "node:fs/promises";
import path from "node:path";
import { COMPLETION_STATUS_TEXT } from "../../core/statuses.js";

/**
 * Maps OpenCode agent names to primary agent roles.
 * @type {Record<string, string>}
 */
const PERSONA_MAPPING = {
  "atlas": "Atlas",
  "orion": "Orion",
  "sage": "Sage",
  "nexus": "Nexus",
  "astra": "Astra",
  "sage-validation": "Sage",
  "orion-release": "Orion",
  "orion-deploy": "Orion",
  "nexus-discovery": "Nexus",
  "phoenix": "Phoenix"
};

/**
 * Gemini Adapter - Transforms OpenCode .md to Gemini Skill format and manages .gemini/ integration.
 */
export class GeminiAdapter {
  constructor({ cwd }) {
    this.cwd = cwd;
  }

  /**
   * Transforms an OpenCode agent or specialized skill file to a Gemini Skill.
   * @param {string} filePath - Path to the source .md file.
   * @returns {Promise<{content: string, name: string}>}
   */
  async transform(filePath) {
    const content = await fs.readFile(filePath, "utf8");
    const fileName = path.basename(filePath, ".md") === "SKILL" 
      ? path.basename(path.dirname(filePath)) 
      : path.basename(filePath, ".md");
      
    const persona = PERSONA_MAPPING[fileName.toLowerCase()] || "Astra";
    
    // Check if already has frontmatter
    if (content.trim().startsWith("---")) {
      return { content, name: fileName };
    }

    // Extract description from the first paragraph after the title
    const descriptionMatch = content.match(/^# .*\n\n(.*)/m);
    const description = descriptionMatch ? descriptionMatch[1].trim() : `Specialized agent for ${fileName}.`;

    const skillContent = `---
name: ${fileName}
description: ${description}
---
# ${persona} Persona

${content}

##  Completion Contract (Mandatory)
Every task completion MUST provide the following payload:
1. **Status**: ${COMPLETION_STATUS_TEXT}
2. **Scope reviewed**: Brief summary of what was analyzed.
3. **Findings**: Evidence-based discoveries.
4. **Files affected**: List of all concrete paths modified or created.
5. **Validation/evidence**: Commands run and their results.
6. **Risks/notes**: Any caveats or technical debt introduced.
7. **Required manager action**: Clear next step for the orchestrator.
8. **Can manager continue**: [yes/no]
`;

    return {
      content: skillContent,
      name: fileName
    };
  }

  /**
   * Deploys transformed skills and agents to their respective directories.
   * Also orchestrates the .gemini/ native integration.
   */
  async deploy(transformedItems, installRoot = ".ai-workflow") {
    const agentsDir = path.join(this.cwd, installRoot, "opencode/agents");
    const skillsDir = path.join(this.cwd, installRoot, "opencode/skills");
    const geminiDir = path.join(this.cwd, ".gemini");
    
    await fs.mkdir(agentsDir, { recursive: true });
    await fs.mkdir(skillsDir, { recursive: true });

    for (const item of transformedItems) {
      const isAgent = PERSONA_MAPPING[item.name.toLowerCase()];
      const targetDir = isAgent ? agentsDir : path.join(skillsDir, item.name);
      const fileName = isAgent ? `${item.name}.md` : "SKILL.md";
      
      const absoluteTargetDir = isAgent ? agentsDir : targetDir;
      await fs.mkdir(absoluteTargetDir, { recursive: true });
      await fs.writeFile(path.join(absoluteTargetDir, fileName), item.content);
    }
    
    // Deploy native Gemini CLI extensions (.gemini/)
    await this.deployNativeExtensions(installRoot);

    // Also deploy root GEMINI.md if it doesn't exist
    await this.deployInstructions();
  }

  /**
   * Orchestrates the .gemini/ directory with copies of .ai-workflow/ assets.
   * Using hard copies instead of symlinks ensures native discovery across all platforms.
   */
  async deployNativeExtensions(installRoot = ".ai-workflow") {
    const geminiDir = path.join(this.cwd, ".gemini");
    const geminiAgents = path.join(geminiDir, "agents");
    const geminiSkills = path.join(geminiDir, "skills");
    const geminiCommands = path.join(geminiDir, "commands");

    await fs.mkdir(geminiAgents, { recursive: true });
    await fs.mkdir(geminiSkills, { recursive: true });
    await fs.mkdir(geminiCommands, { recursive: true });

    // 1. Copy Agents
    const sourceAgentsDir = path.join(this.cwd, installRoot, "opencode/agents");
    const agentFiles = await fs.readdir(sourceAgentsDir).catch(() => []);
    for (const file of agentFiles) {
      if (!file.endsWith(".md")) continue;
      const content = await fs.readFile(path.join(sourceAgentsDir, file), "utf8");
      await fs.writeFile(path.join(geminiAgents, file), content);
    }

    // 2. Copy Skills
    const sourceSkillsDir = path.join(this.cwd, installRoot, "opencode/skills");
    const skillFolders = await fs.readdir(sourceSkillsDir).catch(() => []);
    for (const folder of skillFolders) {
      const skillSource = path.join(sourceSkillsDir, folder, "SKILL.md");
      if (await this.exists(skillSource)) {
        await fs.mkdir(path.join(geminiSkills, folder), { recursive: true });
        const content = await fs.readFile(skillSource, "utf8");
        await fs.writeFile(path.join(geminiSkills, folder, "SKILL.md"), content);
      }
    }

    // 3. Copy Commands (from dist-assets/commands)
    const sourceCommandsDir = path.join(this.cwd, installRoot, "opencode/commands");
    if (await this.exists(sourceCommandsDir)) {
      const commandFiles = await fs.readdir(sourceCommandsDir).catch(() => []);
      for (const file of commandFiles) {
        if (!file.endsWith(".md") && !file.endsWith(".toml")) continue;
        const content = await fs.readFile(path.join(sourceCommandsDir, file), "utf8");
        await fs.writeFile(path.join(geminiCommands, file), content);
      }
    }

    // 4. Inject project settings.json
    const settingsPath = path.join(geminiDir, "settings.json");
    if (!(await this.exists(settingsPath))) {
      const settings = {
        model: { name: "auto" },
        ui: { theme: "Ayu" }
      };
      await fs.writeFile(settingsPath, JSON.stringify(settings, null, 2));
    }
  }

  async exists(p) {
    try {
      await fs.access(p);
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Deploys the main GEMINI.md and .geminiignore instructions to the project root.
   */
  async deployInstructions() {
    const targetPath = path.join(this.cwd, "GEMINI.md");
    const ignorePath = path.join(this.cwd, ".geminiignore");
    
    // Deploy GEMINI.md if it doesn't exist
    try {
      await fs.access(targetPath);
    } catch {
      const instructions = `# AI Workflow Kit - Engineering Governance

## Mandate: Atlas Authority Protocol
All architectural changes must align with the specifications issued by the **Specification Authority** (./specs).

## Ternary Architecture
- **Root:** Orchestration and config.
- **.ai-workflow/:** Implementation (The Muscle).
- **./specs:** Specifications (The Brain).

## Branch Gates
- **main Branch:** READ-ONLY.
- **Validation:** Merges require successful observed validation; full/release workflows may persist EVIDENCE.json.

## Primary Agents
- **Atlas (Orchestrator)**
- **Orion (Strategist)**
- **Sage (Auditor)**
- **Nexus (Spec Architect)**
- **Astra (Developer)**
- **Phoenix (Healer)**

## Operational Preferences
- Use \`token-economy\` and \`minimal-context\` skills.
- **Native Discovery**: This project uses workspace-specific extensions in \`.gemini/\`.
- **Workflow Entry**: Use \`/atlas\` to begin any task.
`;
      await fs.writeFile(targetPath, instructions);
    }

    // Deploy .geminiignore if it doesn't exist
    try {
      await fs.access(ignorePath);
    } catch {
      const ignoreContent = `.ai-workflow/
.ai-workflow-backups/
node_modules/
.git/
EVIDENCE.json
*.tgz
*.zip
*.tar
.gemini/tmp/
.gemini/history/
`;
      await fs.writeFile(ignorePath, ignoreContent);
    }
  }
}
