// @williambeto/ai-workflow/src/core/gates/branch-gate.js
import { execSync } from "node:child_process";
import fs from "node:fs";
import path from "node:path";

function slugify(value = "task") {
  return String(value)
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "")
    .slice(0, 48) || "task";
}

/**
 * BranchGate - Protects main/master and can recover safely onto a scoped branch.
 */
export class BranchGate {
  constructor({ protectedBranches = ["main", "master"], memoryDir, cwd = process.cwd() } = {}) {
    this.protectedBranches = protectedBranches;
    this.logPath = memoryDir ? path.join(memoryDir, "GATE_ALERTS.log") : null;
    this.cwd = cwd;
  }

  run(command) {
    return execSync(command, {
      cwd: this.cwd,
      encoding: "utf8",
      stdio: ["ignore", "pipe", "pipe"]
    }).trim();
  }

  log(event) {
    if (!this.logPath) return;
    const timestamp = new Date().toISOString();
    const logEntry = `[${timestamp}] ${event}\n`;
    try {
      if (!fs.existsSync(path.dirname(this.logPath))) fs.mkdirSync(path.dirname(this.logPath), { recursive: true });
      fs.appendFileSync(this.logPath, logEntry);
    } catch (error) {
      console.error(`Failed to write to gate log: ${error.message}`);
    }
  }

  getDirtyState() {
    const lines = this.run("git status --short").split("\n").filter(Boolean);
    const tracked = lines.filter((line) => !line.startsWith("??"));
    const untracked = lines.filter((line) => line.startsWith("??"));
    return { clean: lines.length === 0, tracked, untracked, lines };
  }

  createScopedBranch(taskSlug) {
    const base = `feat/${slugify(taskSlug)}`;
    let candidate = base;
    let suffix = 2;
    while (true) {
      try {
        this.run(`git show-ref --verify --quiet refs/heads/${candidate}`);
        candidate = `${base}-${suffix++}`;
      } catch {
        break;
      }
    }
    this.run(`git switch -c ${candidate}`);
    return candidate;
  }

  /**
   * @param {string} override
   * @param {{autoRecover?: boolean, taskSlug?: string}} options
   */
  check(override = "", { autoRecover = false, taskSlug = "implementation" } = {}) {
    try {
      const currentBranch = this.run("git rev-parse --abbrev-ref HEAD");
      const isProtected = this.protectedBranches.includes(currentBranch);
      if (!isProtected) return { blocked: false, branch: currentBranch, recovered: false };

      if (override && override.trim().length >= 20) {
        this.log(`AUTHORIZED BYPASS on '${currentBranch}': ${override.trim()}`);
        return { blocked: false, branch: currentBranch, authorized: true, recovered: false };
      }

      if (autoRecover) {
        const dirty = this.getDirtyState();
        if (dirty.tracked.length > 0) {
          const reason = `Unsafe tracked changes prevent Branch Gate Auto-Recovery: ${dirty.tracked.join(", ")}`;
          this.log(`BLOCKED AUTO-RECOVERY on '${currentBranch}': ${reason}`);
          return { blocked: true, branch: currentBranch, reason, dirtyState: dirty };
        }

        const recoveredBranch = this.createScopedBranch(taskSlug);
        this.log(`AUTO-RECOVERED '${currentBranch}' -> '${recoveredBranch}'`);
        return {
          blocked: false,
          branch: recoveredBranch,
          branchBefore: currentBranch,
          recovered: true,
          dirtyState: dirty
        };
      }

      const reason = override && override.trim().length < 20
        ? "Override justification too short (min 20 chars)."
        : `Direct commits to '${currentBranch}' are prohibited.`;
      this.log(`BLOCKED ATTEMPT on '${currentBranch}': ${reason}`);
      return {
        blocked: true,
        branch: currentBranch,
        reason: `${reason} Enable safe auto-recovery or use AI_OVERRIDE with a concrete justification.`
      };
    } catch (error) {
      return { blocked: false, branch: "unknown", error: error.message, recovered: false };
    }
  }
}
